patroni/validator.py

```diff
@@ -0,0 +1,358 @@
+                ConfigParseError("Port {} is already in use.".format(port))
+            elif connect:
+                ConfigParseError("{} is not reachable".format(host_port))
+        except socket.gaierror as e:
+            raise ConfigParseError(e)
+        finally:
+            s.close()
+    return True
+
+
+def comma_separated_host_port(string):
+    if not isinstance(string, string_types):
+        raise ConfigParseError("is not a string")
+    assert all([validate_host_port(s.strip()) for s in string.split(",")]), "didn't pass the validation"
+    return True
+
+
+def validate_host_port_listen(host_port):
+    return validate_host_port(host_port, listen=True)
+
+
+def is_ipv4_address(ip):
+    assert socket.inet_aton(ip)
+    return True
+
+
+def is_ipv6_address(ip):
+    assert socket.inet_pton(socket.AF_INET6, ip)
+    return True
+
+
+def get_major_version(bin_dir=None):
+    if not bin_dir:
+        binary = 'postgres'
+    else:
+        binary = os.path.join(bin_dir, 'postgres')
+    version = subprocess.check_output([binary, '--version']).decode()
+    version = re.match(r'^[^\s]+ [^\s]+ (\d+)(\.(\d+))?', version)
+    return '.'.join([version.group(1), version.group(3)]) if int(version.group(1)) < 10 else version.group(1)
+
+
+def validate_data_dir(data_dir):
+    if not data_dir:
+        raise ConfigParseError("is an empty string")
+    elif os.path.exists(data_dir) and not os.path.isdir(data_dir):
+        raise ConfigParseError("is not a directory")
+    elif not data_directory_empty(data_dir):
+        if not os.path.exists(os.path.join(data_dir, "PG_VERSION")):
+            raise ConfigParseError("doesn't look like a valid data directory")
+        else:
+            with open(os.path.join(data_dir, "PG_VERSION"), "r") as version:
+                pgversion = int(version.read())
+            bin_dir = schema.data.get("postgresql", {}).get("bin_dir", None)
+            if str(pgversion) != get_major_version(bin_dir):
+                raise ConfigParseError("data_dir directory postgresql version doesn't match"
+                                       "with 'postgres --version' output")
+            waldir = ("pg_wal" if pgversion >= 10 else "pg_xlog")
+            if not os.path.isdir(os.path.join(data_dir, waldir)):
+                raise ConfigParseError("data dir for the cluster is not empty, but doesn't contain"
+                                       " \"{}\" directory".format(waldir))
+    return True
+
+
+class Result(object):
+    def __init__(self, status, error="didn't pass validation", path="", data=""):
+        self.status = status
+        self.path = path
+        self.data = data
+        self._error = error
+        if not self.status:
+            self.error = error
+        else:
+            self.error = None
+
+    def __repr__(self):
+        return self.path + (" " + str(self.data) + " " + self._error if self.error else "")
+
+    def __bool__(self):
+        return self.status
```

Comment:
Seems to be unused, because the [status](https://github.com/zalando/patroni/pull/1314/files#diff-05020fb6b44e1d780d1bb03b5b0e257eR164) is checked explicitly.

Reply:
Yes, without checking status failed in python2.7

Reply:
I mean the whole `__bool__()` method is unused and could be dropped.

Reply:
I understood that and removed the method